### PR TITLE
Laboratory work #1, Pershina Olga - 21FPL1

### PR DESCRIPTION
**Pull request by OlyaP12**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -66,7 +106,15 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(frequencies, dict) and isinstance(top, int):
+        slovar_filtered = sorted(slovar.items(), key=itemgetter(1), reverse=True)[:top]
+        print(list(slovar_filtered))
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:111]: remove print

File: lab_1_keywords_tfidf/start.py [hunk 2]
```diff
@@ -34,3 +34,11 @@
     RESULT = None
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
     assert RESULT, 'Keywords are not extracted'
+
+print(main.clean_and_tokenize(target_text))
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/start.py:38]: all this calls must be in `main` statement

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -17,7 +20,15 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
...
     """
-    pass
+
+    if not isinstance(text, str):
+        return None
+    for i in punctuation:
+        text = text.replace(i, "")
+        text = text.lower().strip()
+    tokens = text.split()
+    print(tokens)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:30]: remove

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -33,7 +44,14 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not isinstance(tokens, (str, list)) or not isinstance(stop_words, (list, str)):
+        return None
+    clean_tokens = []
+    for i in tokens:
+        if i not in stop_words:
+            clean_tokens.append(i)
+    print(clean_tokens)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:53]: and this remove too

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -48,7 +66,19 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(tokens, list) and len(tokens) != 0:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:69]: reverse the statement and remove `None` immediately

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -48,7 +66,19 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
...
+        dictionary = {}
+        for word in tokens:
+            if isinstance(word, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:72]: reverse the statement and remove None immediately

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -82,7 +123,19 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
...
+            and all(isinstance(word, str) for word in frequencies.keys())
+            and all(isinstance(freq, int) for freq in frequencies.values())):
+        return None
+    tf_dict = {}
+    number_of_occurrences = sum(frequencies.values())
+    for word, quantity in frequencies.items():
+        if not isinstance(word, str) and not isinstance(quantity, int):
+            return None
+        tf_dict[word] = quantity / number_of_occurrences
+    print(tf_dict)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:137]: remove

File: lab_1_keywords_tfidf/main.py [hunk 8]
```diff
@@ -137,13 +220,25 @@ def calculate_chi_values(expected: dict[str, float], observed: dict[str, int]) -
...
+    for word, value in expected.items():
+        if not isinstance(word, str) or not isinstance(value, float):
+            return None
+    for word, value in observed.items():
+        if not isinstance(word, str) or not isinstance(value, int):
+            return None
+    chi_values = {}
+    for word, value in expected.items():
+        chi_values[word] = ((observed.get(word, 0) - value) ** 2) / value
+    print(chi_values)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:234]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_keywords_tfidf/main.py (hunk 3 above)
Comment (OlyaP12) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (OlyaP12) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 6 above)
Comment (OlyaP12) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 8 above)
Comment (OlyaP12) [reply lab_1_keywords_tfidf/main.py]: all 'print' removed

**Pull request closed without merging**